### PR TITLE
BKNDLSS-28731 Rating

### PR DESCRIPTION
**Pull request by Saharij**

### CODE REVIEW

File: components/bl-rating-component/styles/index.less [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// you can use LESS variables from the current Theme and app extensions of the UI-Container
+@bl-customComponent-badge-themeColor: @themePrimary;
```
Comment (v-excelsior) [inline components/bl-rating-component/styles/index.less:2]: Could we add variables for disabled icons?

File: components/bl-rating-component/src/RatingIcon.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import { icons } from './icons.js'
+
+const halfIconStyle = { width: '50%', overflow: 'hidden', position: 'absolute' }
+
+const RatingIcon = ({
```
[the remark this answers is not in the document]
Comment (v-excelsior) [reply components/bl-rating-component/src/RatingIcon.js:5]: Could we rename it to RatingItem and split into smaller components?
RatingIcon
RatingLabel

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-rating-component/styles/index.less (hunk 1 above)
Comment (Saharij) [reply components/bl-rating-component/styles/index.less:2]: Sure

File: components/bl-rating-component/src/index.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    setHoverValue(roundValueToPrecision(iconsAmount * percent + precision / 2, precision).toString());
+  };
+
+  const handleMouseLeave = () => {
+    if (disabled) return;
+
+    setHoverValue();
+  };
+
+  const classes = useMemo(() => {
```
Comment (v-excelsior) [inline components/bl-rating-component/src/index.js:67]: create an hook useRatingClassList

File: components/bl-rating-component/src/index.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    const arr = ['bl-customComponent-rating', ...classList];
```
Comment (v-excelsior) [inline components/bl-rating-component/src/index.js:68]: empty line

File: components/bl-rating-component/src/rating-icon.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  disabled,
+  ratingValue,
+  onChange
+}) => {
+  const halfInputValue = index + 0.5;
+  const inputValue = index + 1;
+
+  return (
+    <div className="bl-customComponent-ratingIcon rating-icon">
+      <label
```
Comment (v-excelsior) [inline components/bl-rating-component/src/rating-icon.js:18]: Could we split it to
`HalfPointControll` and `FullPointControll` components with `Label` and `Input` components inside?

File: components/bl-rating-component/src/rating-icon.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            <path d={ratingValue >= inputValue ? icons[`${icon}Active`] : icons[icon]} />
+          </svg>
+        </span>
+      </label>
+      <input
+        type="radio"
+        name="rating"
+        value={inputValue}
+        id={inputValue}
+        disabled={disabled}
```
Comment (v-excelsior) [inline components/bl-rating-component/src/rating-icon.js:62]: Did you apply `editorconfig`?

File: components/bl-rating-component/styles/index.less [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    pointer-events: none;
+    cursor: default;
+  }
+}
+
+.bl-customComponent-rating__icons {
+  display: inline-flex;
+}
+
+.rating-icon {
```
Comment (v-excelsior) [inline components/bl-rating-component/styles/index.less:26]: all classes without customComponent part should be inside *-customComponent-* class

```
.foo-customComponent-bar {
 .any{}
}
```

File: components/bl-rating-component/src/rating-icon.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
+import { icons } from './icons.js';
+
+const halfIconStyle = { width: `50%`, overflow: 'hidden', position: 'absolute' };
```
Comment (v-excelsior) [inline components/bl-rating-component/src/rating-icon.js:3]: could we move it to less, i think if we add one more class for half icon that will be possible

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-rating-component/src/assets/icons.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+export const iconsMap = {
```
Comment (v-excelsior) [inline components/bl-rating-component/src/assets/icons.js:1]: move to helpers

File: components/bl-rating-component/src/hooks/useRatingClassList.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import { useMemo } from 'react';
+
+export const useRatingClassList = (disabled, classList) => {
```
Comment (v-excelsior) [inline components/bl-rating-component/src/hooks/useRatingClassList.js:3]: rename file to kebab-case and move to helpers

File: components/bl-rating-component/src/index.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+import { useRatingClassList } from './hooks/useRatingClassList';
+import { iconParams } from './utils/icon-params';
+import { iconsList } from './utils/icons-list.js';
+import { roundValueToPrecision } from './utils/number';
+import { RatingIcon } from './rating-icon';
+
+const RATING_PRECISION = 0.5;
+
+export default function RatingComponent({
```
Comment (v-excelsior) [inline components/bl-rating-component/src/index.js:11]: in one line 

`export default function RatingComponent({ component, eventHandlers }) {`

File: components/bl-rating-component/src/index.js [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+  const classes = useRatingClassList(disabled, classList);
+
+  const rootRef = useRef();
+  const [ratingValue, setRatingValue] = useState(defaultValue);
+  const [hoverValue, setHoverValue] = useState();
+  const icons = iconsList(iconsAmount);
+
+  const value = hoverValue || ratingValue;
+
+  if (!display) {
```
Comment (v-excelsior) [inline components/bl-rating-component/src/index.js:33]: return null always must be after all hooks calling

https://bobbyhadz.com/blog/react-rendered-more-hooks-than-during-previous-render

File: components/bl-rating-component/src/index.js [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  const handleMouseMove = event => {
+    if (disabled) {
+      return;
+    }
+
+    const [left, width] = iconParams(rootRef.current);
+
+    const percent = (event.clientX - left) / (width * iconsAmount);
+
+    setHoverValue(roundValueToPrecision(iconsAmount * percent + RATING_PRECISION / 2, RATING_PRECISION).toString());
```
Comment (v-excelsior) [inline components/bl-rating-component/src/index.js:63]: create a function `getIconHoverValue(rootRef, iconsAmount)`

File: components/bl-rating-component/src/index.js [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+  };
+
+  const handleMouseLeave = () => {
+    if (disabled) {
+      return;
+    }
+
+    setHoverValue();
```
Comment (v-excelsior) [inline components/bl-rating-component/src/index.js:71]: setHoverValue(null)

I think it's good practice - to set null, when we zeroing value by hands

File: components/bl-rating-component/src/rating-icon/components/full-point-controll.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
+export const FullPointControll = ({
```
Comment (v-excelsior) [inline components/bl-rating-component/src/rating-icon/components/full-point-controll.js:1]: in one line

File: components/bl-rating-component/src/rating-icon/components/half-point-controll.js [hunk 16]
```diff
@@ -0,0 +1,37 @@
+export const HalfPointControll = ({
```
Comment (v-excelsior) [inline components/bl-rating-component/src/rating-icon/components/half-point-controll.js:1]: in one line

File: components/bl-rating-component/src/rating-icon/index.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
+import { iconsMap } from '../assets/icons';
+import { FullPointControll, HalfPointControll } from './components';
+
+export const RatingIcon = ({
```
Comment (v-excelsior) [inline components/bl-rating-component/src/rating-icon/index.js:4]: in one line

File: components/bl-rating-component/src/utils/number.js [hunk 18]
```diff
@@ -0,0 +1,13 @@
+function getDecimalPrecision(num) {
```
Comment (v-excelsior) [inline components/bl-rating-component/src/utils/number.js:1]: please, choose the named or anonimous function and try to use them consistently

File: components/bl-rating-component/src/utils/number.js [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+  const decimalPart = num.toString().split('.')[1];
+  
+  return decimalPart ? decimalPart.length : 0;
+}
+
+export const roundValueToPrecision = (value, precision) => {
+  if (value == null) return value;
```
Comment (v-excelsior) [inline components/bl-rating-component/src/utils/number.js:8]:
```
if (Object.is(value, null)) {
  return value;
}
```

Also, use strict evaluation everywhere(===)

File: components/bl-rating-component/src/utils/number.js (hunk 18 above)
Comment (v-excelsior) [inline components/bl-rating-component/src/utils/number.js:1]: move all utils to helpers dir

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-rating-component/component.json [hunk 20]
```diff
@@ -0,0 +1,88 @@
+{
```
Comment (v-excelsior) [inline components/bl-rating-component/component.json:1]: We need a property to enable/disable half rating changing

File: components/bl-rating-component/src/index.js [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+
+  if (!display) {
+    return null;
+  }
+
+  const handleRatingValue = useCallback(({ target: { value  } }) => {
+    setRatingValue(prevState => prevState === value ? 0: value);
+
+    if (onRatingChange) {
+      onRatingChange({ curentRating: value });
```
Comment (v-excelsior) [inline components/bl-rating-component/src/index.js:41]: Why we call handler 2 times in some cases? I think that should works in other way. Create object with arguments and pass it one time.

File: components/bl-rating-component/src/index.js [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+  };
+
+  return (
+    <div className={ classes }>
+      <div
+        ref={ rootRef }
+        onMouseMove={ handleMouseMove }
+        onMouseLeave={ handleMouseLeave }
+        className="bl-customComponent-rating__icons">
+        {icons.map((_, index) => (
```
Comment (v-excelsior) [inline components/bl-rating-component/src/index.js:81]: My editor config adds space after `{`. Please check whether you apply that one that I shared with you before.

File: components/bl-rating-component/src/rating-icon/index.js [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+  ratingValue,
+  onChange,
+}) => {
+  const halfInputValue = index + 0.5;
+  const inputValue = index + 1;
+
+  return (
+    <div className="rating-icon">
+      <HalfPointControll
+        icon={ ratingValue >= halfInputValue ? iconsMap[`${icon}Active`] : iconsMap[icon] }
```
Comment (v-excelsior) [inline components/bl-rating-component/src/rating-icon/index.js:18]: utility function `getIcon`?

File: components/bl-rating-component/src/utils/icon-params.js [hunk 24]
```diff
@@ -0,0 +1,6 @@
+export const iconParams = rootNode => {
```
Comment (v-excelsior) [inline components/bl-rating-component/src/utils/icon-params.js:1]: getIconsParans

File: components/bl-rating-component/src/utils/icons-list.js [hunk 25]
```diff
@@ -0,0 +1 @@
+export const iconsList = iconsAmount => new Array(iconsAmount).fill(1);
```
Comment (v-excelsior) [inline components/bl-rating-component/src/utils/icons-list.js:1]: please, don't create file for one row function

File: components/bl-rating-component/README.md [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+<dt>Icon</dt>
+<dd>Each icon has its own rating, when you click on it, the rating will change depending on the value of a particular icon.</dd>
+</dl>
+
+### Component Properties
+
+  Name              | Default value     | Description
+ -------------------|-------------------|-------------------------------------------------------------------------------------
+  Disabled          | false             | If you need the rating to be disabled, you can click on the `Disabled` checkbox.
+  Icon              | Star              | You can select the icon to be displayed.
```
Comment (v-excelsior) [inline components/bl-rating-component/README.md:20]: Default value is "star"(add quotes as it string), Star is label. Also for select property type would be helpful to write all possible values at description.

File: components/bl-rating-component/component.json [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+          "label": "Min Rating"
+        },
+        {
+          "id": "maxRating",
+          "label": "Max Rating"
+        }
+      ]
+    }
+  ],
+  "actions": []
```
Comment (v-excelsior) [inline components/bl-rating-component/component.json:106]: remove it

File: components/bl-rating-component/src/helpers/use-rating-class-list.js [hunk 28]
```diff
@@ -0,0 +1,15 @@
...
+    
+    if (disabled) {
+      listClassesOf.push('disabled');
+    }
+    
+    return listClassesOf.join(' ');
+  }, [classList, disabled]);
+  
+  return classes;
+};
```
Comment (v-excelsior) [inline components/bl-rating-component/src/helpers/use-rating-class-list.js:15]:
```
  export const useRatingClassList = (disabled, classList) => useMemo(() => {
    const listClassesOf = ['bl-customComponent-rating', ...classList]

    if (disabled) {
      listClassesOf.push('disabled')
    }

    return listClassesOf.join(' ')
  }, [classList, disabled])
```

File: components/bl-rating-component/src/index.js [hunk 29]
```diff
@@ -0,0 +1,91 @@
...
+  const [hoverValue, setHoverValue] = useState();
+
+  const icons = iconsList(iconsAmount);
+  const value = hoverValue || ratingValue;
+
+  const handleRatingValue = useCallback(({ target: { value  } }) => {
+    setRatingValue(prevState => prevState === value ? 0: value);
+
+    if (onRatingChange) {
+      const requestParams = {
```
Comment (v-excelsior) [inline components/bl-rating-component/src/index.js:33]: requestParams -> context | eventContext | eventData | contextBlocks

File: components/bl-rating-component/src/helpers/get-icon-hover-value.js [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+  
+  return Number(nearest.toFixed(getDecimalPrecision(precision)));
+}
+
+export const getIconHoverValue = (rootNode, iconsAmount, clientX, setHoverValue, precision) => {
+  const { left } = rootNode.getBoundingClientRect();
+  const { width } = rootNode.firstChild.getBoundingClientRect();
+  
+  const ratingPrecision = Number(precision);
+  const percent = (event.clientX - left) / (width * iconsAmount);
```
Comment (v-excelsior) [inline components/bl-rating-component/src/helpers/get-icon-hover-value.js:22]: event.clientX -> clientX?

File: components/bl-rating-component/src/helpers/icons.js [hunk 31]
```diff
@@ -0,0 +1,11 @@
+export const iconsMap = {
+  star: `M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22
+  9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z`,
```
Comment (v-excelsior) [inline components/bl-rating-component/src/helpers/icons.js:3]: Could we separate it by new line?

File: components/bl-rating-component/src/index.js [hunk 32]
```diff
@@ -0,0 +1,91 @@
...
+      onRatingChange(requestParams);
+    }
+  }, []);
+
+  const handleMouseMove = event => {
+    if (disabled) {
+      return;
+    }
+
+    getIconHoverValue(rootRef.current, iconsAmount, event.clientX, setHoverValue, precision);
```
Comment (v-excelsior) [inline components/bl-rating-component/src/index.js:54]: setHoverValue(getIconHoverValue(rootRef.current, iconsAmount, event.clientX, precision))

Also, could we pass the event, and instead `rootRef.current` get currentTarget from event? In this case that could look like below:

`setHoverValue(getIconHoverValue(event, iconsAmount, precision))`

File: components/bl-rating-component/styles/index.less [hunk 33]
```diff
@@ -0,0 +1,63 @@
...
+    }
+  }
+  
+  .half-icon {
+    position: absolute;
+    width: 50%;
+    overflow: hidden;
+  }
+}
+
```
Comment (v-excelsior) [inline components/bl-rating-component/styles/index.less:58]: remove unneeded empty lines

File: components/bl-rating-component/src/rating-icon/index.js [hunk 34]
```diff
@@ -0,0 +1,45 @@
+import { iconsMap } from '../helpers/icons';
+import { FullPointControll } from './full-point-controll';
+import { HalfPointControll } from './half-point-controll';
+
+const getIcon = (ratingValue, inputValue, icon) => {
+  return ratingValue >= inputValue ? iconsMap[`${icon}Active`] : iconsMap[icon];
+};
+
+export const RatingIcon = props => {
+  const {
```
Comment (v-excelsior) [inline components/bl-rating-component/src/rating-icon/index.js:10]: in one line

File: components/bl-rating-component/src/rating-icon/half-point-controll.js [hunk 35]
```diff
@@ -0,0 +1,37 @@
+export const HalfPointControll = ({
```
Comment (v-excelsior) [inline components/bl-rating-component/src/rating-icon/half-point-controll.js:1]: in one line

File: components/bl-rating-component/src/rating-icon/full-point-controll.js [hunk 36]
```diff
@@ -0,0 +1,37 @@
+export const FullPointControll = ({
```
Comment (v-excelsior) [inline components/bl-rating-component/src/rating-icon/full-point-controll.js:1]: in one line

File: components/bl-rating-component/src/index.js [hunk 37]
```diff
@@ -0,0 +1,91 @@
+import { useCallback, useRef, useState } from 'react';
+
+import { getIconHoverValue, useRatingClassList } from './helpers';
+import { RatingIcon } from './rating-icon';
+
+const iconsList = iconsAmount => new Array(iconsAmount).fill(1);
+
+export default function RatingComponent({ component, eventHandlers }) {
+  const {
```
Comment (v-excelsior) [inline components/bl-rating-component/src/index.js:9]: in one line

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-rating-component/src/helpers/use-rating-class-list.js [hunk 38]
```diff
@@ -0,0 +1,11 @@
+import { useMemo } from 'react';
+
+export const useRatingClassList = (disabled, classList) => useMemo(() => {
```
Comment (v-excelsior) [inline components/bl-rating-component/src/helpers/use-rating-class-list.js:3]: Could we change the order?

`classList, disabled`

The idea is first argument must be most important

**Review by v-excelsior — CHANGES_REQUESTED** [review]

**Pull request merged**